---
title: "Workflow_enrichment"
author: "Sandrini Giada"
date: "2025-02-20"
output: html_document

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(
echo = TRUE,
message = FALSE,
warning = FALSE
)
```

## 1 Introduction

This R Markdown file provides a complete workflow for performing pathway enrichment analysis to improve the biological interpretability of transcriptomic results. The workflow allows grouping pathways into predefined biological categories and visualizing multiple comparisons simultaneously, providing a global view of biological alterations.

**Example dataset:** Transcriptomic data from prostate cancer studies have been used here. The differential expression (DE) comparisons considered are:

- PRIMARY vs NORMAL  

- CRPC vs NORMAL  

- NEPC vs NORMAL  

These comparisons were also reported in *Cacciatore et al.*, *Civenni et al.*, *Pardella et al.*


The workflow uses pre-processed gene sets from curated databases, including:

- Hallmark

- MitoCarta

- KEGG

- Reactome

Enrichment analyses included in this workflow are:

- GSEA (Gene Set Enrichment Analysis)

- camera

- Over-Representation Analysis (ORA)

The workflow also uses custom functions to generate publication-ready plots, which are automatically saved into dedicated folders. This setup allows easy comparison across multiple experimental conditions. For example, when comparing two treatments against a control, it enables visualization of pathway enrichment for both conditions in a single plot. This approach provides a broader understanding of altered cellular mechanisms beyond just the most deregulated pathways.
This R Markdown file provides a complete workflow for performing pathway enrichment analysis using multiple gene sets from curated databases, including:

## 2 Environment Setup and Pathway Loading

#### 2.1 Loading packages and user functions

```{r}
library("limma")
library("msigdbr")
library("plyr")
library("dplyr")
library("ggplot2")
library("forcats")
library("biomaRt")

source("Enrichment_analysis_functions.R")
```

#### 2.2 Load pre-processed pathway object

The object Pathways_4Enrichment.rds contains:
sorted_pathways: pathways grouped by category (Hallmarks, Reactome, KEGGâ€¦)
GSEA_combined: gene sets formatted for GSEA and camera
If you prefer regenerating it, the code is provided but commented.

Optionally regenerate the pathway object:
```{r}
#Pathways_4Enrichment = Prepare_Enrichment(my_species = "Homo sapiens",
#                            sorted_pathway_directory = "Sorted_pathways")
#saveRDS(Pathways_4Enrichment, file = "Pathways_4Enrichment_human.rds")
#Pathways_4Enrichment = Prepare_Enrichment(my_species = "Mus musculus",
#                            sorted_pathway_directory = "Sorted_pathways")
#saveRDS(Pathways_4Enrichment, file = "Pathways_4Enrichment_mouse.rds")
```

Load the ready-to-use object

```{r}
Pathways_4Enrichment = readRDS("Pathways_4Enrichment_human.rds")
sorted_pathways = Pathways_4Enrichment$sorted_pathways
GSEA_combined = Pathways_4Enrichment$GSEA_combined
```

#### 2.3 Load Differential Expression Results
The enrichment pipeline accepts pre-computed differential expression (DE) results. By default, it automatically recognizes outputs from DESeq2 and limma packages, including the necessary columns for:

- log2FoldChange (my_log2FC)

- my_p_value

- symbol_weight

- Feature_name

Note: If you are using DE results from other tools, you need to specify which columns correspond to my_log2FC, my_p_value, symbol_weight, Feature_name.

Load the DE results object and create the list that 

```{r}
ANALYSIS_1 = readRDS("Example_dataset/DE_PRIMARYvsNORMAL.rds")[["res.noNA"]]
ANALYSIS_2 = readRDS("Example_dataset/DE_CRPCvsNORMAL.rds")[["res.noNA"]]
ANALYSIS_3 = readRDS("Example_dataset/DE_NEPCvsNORMAL.rds")[["res.noNA"]]
list_DE_outputs = list(FirstComparison = ANALYSIS_1,
                       SecondComparison = ANALYSIS_2,
                       ThirdComparison = ANALYSIS_3)
```

## 3 Run Enrichment Analysis

#### GSEA

```{r, results='hide'}
Enrich_res = Compute_Enrichment(list_DE_outputs,
                                my_GeneSet = GSEA_combined,
                                my_sorted_pathways = sorted_pathways,
                                enrich_method = "gsea")
Plot_gsea = Plot_Enrichment(Enrich_res, fold_name = "Enrichment_fig_gsea")
```

Plot Reactome - Adaptive immunity, as an example

```{r fig.dim=c(8,8), out.width='50%', fig.align='center'}
Plot_gsea[["reactome_immune_adaptive"]]
```

#### camera

```{r, results='hide'}
Enrich_res = Compute_Enrichment(list_DE_outputs,
                                my_GeneSet = GSEA_combined,
                                my_sorted_pathways = sorted_pathways,
                                enrich_method = "camera")
Plot_camera = Plot_Enrichment(Enrich_res, fold_name = "Enrichment_fig_camera")
```

Plot Hallmark results, as an example

```{r fig.dim=c(8,11), out.width='60%', fig.align='center'}
Plot_gsea[["hallmarks"]]
```

#### ORA

```{r, results='hide'}
Enrich_res = Compute_Enrichment(list_DE_outputs,
                                my_GeneSet = GSEA_combined,
                                my_sorted_pathways = sorted_pathways,
                                enrich_method = "ORA")
Plot_ORA_up = Plot_Enrichment(Enrich_res$UP, fold_name = "Enrichment_fig_ORA_up")
Plot_ORA_down = Plot_Enrichment(Enrich_res$DOWN, fold_name = "Enrichment_fig_ORA_down")
```

Plot Reactome - Cell Damage, as an example

```{r fig.dim=c(10,11), out.width='60%', fig.align='center'}
Plot_ORA_up[["reactome_damage"]]
```

Plot Reactome - Innate Immunity

```{r fig.dim=c(8,11), out.width='60%', fig.align='center'}
Plot_ORA_down[["reactome_immune_innate"]]
```

